qiskit/primitives/backend_estimator_dev.py

```diff
@@ -0,0 +1,612 @@
+        for cbit, qubit in enumerate(measured_qubit_indices):
+            if pauli.x[qubit]:
+                if pauli.z[qubit]:
+                    circuit.sdg(qubit)
+                circuit.h(qubit)
+            circuit.measure(qubit, cbit)
+        return circuit
+
+    ################################################################################
+    ## COMPOSITION
+    ################################################################################
+    def _compose_measurements(
+        self,
+        base: QuantumCircuit,
+        measurements: Sequence[QuantumCircuit] | QuantumCircuit,
+    ) -> tuple[QuantumCircuit, ...]:
+        """Compose measurement circuits with base circuit considering final layout."""
+        if isinstance(measurements, QuantumCircuit):
+            measurements = (measurements,)
+        return tuple(self._compose_single_measurement(base, meas) for meas in measurements)
+
+    def _compose_single_measurement(
+        self, base: QuantumCircuit, measurement: QuantumCircuit
+    ) -> QuantumCircuit:
+        """Compose single measurement circuit with base circuit considering final layout.
+
+        Args:
+            base: a quantum circuit with final_layout entry in its metadata
+            measurement: a quantum circuit with
+
+        Returns:
+            A compsite quantum circuit
+        """
+        layout = base.metadata.get("final_layout")
+        transpile_options = {**self.transpile_options.__dict__}  # TODO: avoid multiple copies
+        transpile_options.update({"initial_layout": layout})
+        transpiled_measurement = transpile(measurement, self.backend, **transpile_options)
+        circuit = base.compose(transpiled_measurement)
+        circuit.metadata = {**base.metadata, **measurement.metadata}
+        circuit.metadata.pop("measured_qubit_indices")  # TODO: `measured_qubits`
+        return circuit
+
+    ################################################################################
+    ## DEPRECATED
+    ################################################################################
+    # Note: to allow `backend` as positional argument while deprecated in place
+    def __new__(  # pylint: disable=signature-differs
+        cls,
+        backend: Backend,  # pylint: disable=unused-argument
+        **kwargs,  # pylint: disable=unused-argument
+    ):
+        self = super().__new__(cls)
+        return self
+
+    def _call(
+        self,
+        circuits: Sequence[int],
+        observables: Sequence[int],
+        parameter_values: Sequence[Sequence[float]],
+        **run_options,
+    ) -> EstimatorResult:
+        raise NotImplementedError("This method has been deprecated, use `run` instead.")
+
+
+################################################################################
+## OBSERVABLE DECOMPOSER
+################################################################################
+class ObservableDecomposer(ABC):
```

Comment:
Why do you need to introduce a new decomposer class instead of just functions? Do you have any idea of other grouping algorithms in addition to Abelian grouping?

Reply:
This can be solved by declaring these classes as [private](https://namingconvention.org/python/), so that we can leverage them internally but don't settle for anything prematurely 🙂